File: PSMusic/PSMusic.Server/Services/Implementations/AuthService.cs
using AutoMapper;
using PSMusic.Server.Models.DTO.Auth;
using PSMusic.Server.Models.DTO.User;
using PSMusic.Server.Models.Entities;
using PSMusic.Server.Repositories.Interfaces;
using PSMusic.Server.Services.Interfaces;
using System.Security.Claims;

namespace PSMusic.Server.Services.Implementations
{
    public class AuthService : IAuthService
    {
        private readonly IConfiguration _configuration;
        private readonly IUserRepository _userRepository;
        private readonly IMapper _mapper;
        private readonly ITokenGenerator _tokenGenerator;

        public AuthService(IConfiguration configuration, IUserRepository userRepository, IMapper mapper, ITokenGenerator tokenGenerator)
        {
            _configuration = configuration;
            _userRepository = userRepository;
            _mapper = mapper;
            _tokenGenerator = tokenGenerator;
        }

        public async Task<AuthResDTO> Login(AuthReqDTO req)
        {
            User? user = await _userRepository.GetUserByUsername(req.Username);
            if (user == null) return new AuthResDTO { IsSuccess = false, Message = "Tài khoản không tồn tại" };

            if (user.VerifyPassword(req.Password))
            {
                var token = _tokenGenerator.GenerateToken(user);
                var refreshToken = _tokenGenerator.GenerateRefreshToken(user);
                return new AuthResDTO { IsSuccess = true, Message = "Đăng nhập thành công", Token = token, RefreshToken = refreshToken }; 
            }
            else return new AuthResDTO { IsSuccess = false, Message = "Sai mật khẩu" };
        }

        public async Task<AuthResDTO> Register(CreateUserDTO user)
        {
            bool emailExistence = await _userRepository.CheckEmailExistence(user.Email);
            if (emailExistence) return new AuthResDTO { IsSuccess = false, Message = "Email đã tồn tại" };

            User? userExist = await _userRepository.GetUserByUsername(user.Username);
            if (userExist != null) return new AuthResDTO { IsSuccess = false, Message = "Tài khoản đã tồn tại" };
            
            User userEntity = _mapper.Map<User>(user);
            userEntity.SetPassword(userEntity.Password);
            if (await _userRepository.Add(userEntity)) return new AuthResDTO { IsSuccess = true, Message = "Đăng ký thành công" };
            else return new AuthResDTO { IsSuccess = false, Message = "Tạo tài khoản thất bại"};
        }

        public async Task<AuthResDTO> Refresh(string refreshToken)
        {
            var principal = _tokenGenerator.ValidateRefreshToken(refreshToken);
            if (principal == null) return new AuthResDTO { IsSuccess = false, Message = "Refresh token không hợp lệ hoặc đã hết hạn" };

            var userId = principal.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value;
            if (userId == null) return new AuthResDTO { IsSuccess = false, Message = "Token không chứa thông tin người dùng" };

            var user = await _userRepository.GetUserById(int.Parse(userId));
            if (user == null) return new AuthResDTO { IsSuccess = false, Message = "Người dùng không tồn tại" };

            var newToken = _tokenGenerator.GenerateToken(user);
            return new AuthResDTO { IsSuccess = true, Message = "Refresh thành công", Token = newToken };
        }
    }
}

File: PSMusic/psmusic.client/src/services/authService.js
import axiosInstance from "./axiosInstance";

const authService = {
    async login(username, password)
    {
        try {
            const res = await axiosInstance.post("/auth/login", {
                username,
                password
            });
            return res.data;
        } catch (error) {
            //console.log(error.response);
            if (error.response?.status === 400)
            {
                return {
                    isSuccess: false,
                    message: error.response.data?.message || "Tên đăng nhập hoặc mật khẩu không đúng"
                };
            }
            //console.error("Login API error:", error);
            return { isSuccess: false, message: "Đã xảy ra lỗi không xác định" };
        }
    },

    async getCurrentUser() {
        try {
            const res = await axiosInstance.get("/auth/me");
            if (res.status === 200 && res.data) return { isSuccess: true, data: res.data };
            throw new Error("Không thể lấy thông tin người dùng");
        } catch (error) {
            if (error.response?.status !== 401) {
                //console.error("Get current user error:", error);
            }

            if (error.response?.status === 401) {
                throw new Error("Chưa đăng nhập hoặc phiên đăng nhập đã hết hạn");
            }

            throw new Error(
                error.response?.data?.message || "Không thể lấy thông tin người dùng"
            );
        }
    },

    async logout() {
        try {
            const res = await axiosInstance.post("/auth/logout");
            if (res.status === 200 && res.data) return res.data;
        } catch (error) {
            //console.error("Logout error:", error);
            return {
                isSuccess: true,
                message: "Đã đăng xuất"
            };
        }
    },

    async register(user) {
        try {
            const res = await axiosInstance.post("/auth/register", user);
            return res.data;
        } catch (error) {
            //console.error("Register error:", error);
            return { 
                isSuccess: false,
                message: error.response.data?.message || "Đăng ký thất bại"
            }
        }
    },

    async refresh() {
        try {
            const res = await axiosInstance.post("/auth/refresh");
            if (res.status === 200 && res.data) return res.data;
        } catch (error) {
            if (error.response?.status !== 401) {
                //console.error("Refresh token error:", error);
            }
            return {
                isSuccess: false,
                message: "Không thể làm mới phiên đăng nhập"
            };
        }
    }
}

export default authService;